README.md

```diff
@@ -97,60 +97,76 @@ let encodedTrifleSignedDataProto = try trifleSignedData.serialize()
 // from backup, ... etc)
 
 // Create cert request
-val certReq = TrifleApi.generateMobileCertificateRequest(entity = entity, keyHandle = keyHandle)
+val certReq: CertificateRequest = TrifleApi.generateMobileCertificateRequest(entity, keyHandle)
 
-// Serialize to proto to be sent over wire
+// Serialize as an opaque ByteArray
+val encoded: ByteArray = certReq.serialize()
 
-// Send certificate request to Certificate Authority endpoint. Response will be [Data]
-// Iterate over each Data to convert to TrifleCertificate
+// Send certificate request to Certificate Authority endpoint. Response will be List<ByteArray>
+val response: List<ByteArray>
 
-// certs is an array of certificates where [0] will be device certificate
+// Iterate over each ByteArray to convert to Certificate
+val certs: List<Certificate> = response.map { Certificate.deserialize(it) }
+
+// certs is a list of certificates where [0] will be device certificate
 // and the rest of the elements will be intermediate chain.
 
 // Check if app has the root cert of Certificate Authority (CA). 
 
 // Validate cert matches the certificate request (so generated key) 
 // and the root (so it has been generated by the right CA).
-// TBD var isValid = TrifleApi.verify(csr, certs, anchor)
+val result: Result<Unit> = certs[0].verify(certReq, certs, root)
+if (result.isFailure) {
+  failureHandler(result)
+} 
+// otherwise, continue
 
-// Once it passes validation, certReq is no longer needed and it can be deleted
+// Once it passes validation, certReq is no longer needed, and it can be deleted
 // Store cert along with the respective keyHandle
         
 // To check only for the validity of a stored cert, you can do either of below choices
 // Option 1 is a more complete check of the device cert and the full chain
-// TBD var isValid = TrifleApi.verify(certs)
+val result: Result<Unit> = cert.verify(ancestorCertificateChain = certs.drop(1))
 
 // option 2 only checks the validity of the device cert
-// TBD var isValid = TrifleApi.verify(certs)
+val result: Result<Unit> = cert.verify()
 
 // Sign the data
-let trifleSignedData = TrifleApi.createSignedData(
-                                        dataThatIsSigned, 
-                                        keyHandle, 
-                                        certs )
+val signedData: SignedData = TrifleApi.createSignedData(clientData, keyHandle, certs)
 
-// Serialize to proto to be sent over wire
-val encodedTrifleSignedDataProto = trifleSignedData.serialize()
+// Serialize as an opaque ByteArray
+val encodedSignedData: ByteArray = trifleSignedData.serialize()
+
+// Deserialize the encoded ByteArray as a SignedData
+val decodedSignedData: SignedData = SignedData.deserialize(encodedSignedData)
+
+// Verify and extract the signed data
+val result: Result<VerifiedData> = decodedSignedData.verifyAndExtract(root)
+val verifiedData: VerifiedData = if (result.isFailure) {
+  failureHandler(result)
+} else {
+  result.getOrThrow()
```

Comment:
is this done on server side? client wouldn't do this, would they?

Reply:
Yea, let me remove that bit